Adding -L option to CoqProject for recursively building and importing dependencies

Durbatuluk1701 opened the pull request:
<!-- Thank you for your contribution.
     Make sure you read the contributing guide and fill this template. -->
The changes I have proposed in the pull request are adding an option to _CoqProjects of the form `-L <physDir> <logname>`.

This option will be interpreted by the coq_makefile to go into `physDir` and build the "library" utilizing a _CoqProject file at the top level of `physDir`. After the building is complete, it will act as if it was a `-R <physDir> <logname>` and make the built files available under the logical name `logname`.

I was not sure if this change was major enough to merit a CEP, but I can try to write one up if we believe it is necessary.


<!-- Remove anything that doesn't apply in the following checklist. -->

<!-- If there is a user-visible change and testing is not prohibitively expensive: -->
- [ ] Added / updated **test-suite**.

<!-- If this is a feature pull request / breaks compatibility: -->
- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.
  <!-- Check if the following applies, otherwise remove these lines. -->
  - [ ] Documented any new / changed **user messages**.
  - [ ] Updated **documented syntax** by running `make doc_gram_rsts`.

<!-- If this breaks external libraries or plugins in CI: -->
- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

gares commented:
I believe you should write some doc, it should help understanding your patch. Please provide an example/use-case in the doc. I'm talking about doc/sphynx.

About the makefile, I have the impression that if the sub-library fails to build, the error is not popped up blocking the build of the main library.

Durbatuluk1701 commented:
Good catch on the makefile @gares, should be fixed now. 
Also pushed some documentation and an initial test case to hopefully make the goal of this PR more clear.

silene commented:
Do we need yet another option? Would not it be sufficient to teach `coq_makefile` that, if it finds a `_CoqProject` file at the target directory of a `-R` or `-Q` option, it should consider this directory as being built independently. In case this check is a bit too frail, there could be an additional check, e.g., the original `_CoqProject` file should not mention any file in the target directory.

gares commented:
Actually this remark makes me think that each subproject is likely to have its own logical root, so I don't get why -L takes one. If there is a mismatch between the logical path given in `-L suproject p1` and the one in `subproject/_CoqProject: -R . p2` I wonder what happens, likely Coq refusing to load .vo files from the subproject.

Durbatuluk1701 commented:
> Do we need yet another option? Would not it be sufficient to teach `coq_makefile` that, if it finds a `_CoqProject` file at the target directory of a `-R` or `-Q` option, it should consider this directory as being built independently. In case this check is a bit too frail, there could be an additional check, e.g., the original `_CoqProject` file should not mention any file in the target directory.

My original thought was to take this approach, however one of the use-cases I have found this modification helpful for is having some sub-project depend on another sub-project that is not a strict sub-directory. (i.e. `-L ../proj1 Proj1`). 
I believe that -R and -Q both require that any loaded packages be a sub-directory. 

> Actually this remark makes me think that each subproject is likely to have its own logical root, so I don't get why -L takes one. If there is a mismatch between the logical path given in -L suproject p1 and the one in subproject/_CoqProject: -R . p2 I wonder what happens, likely Coq refusing to load .vo files from the subproject.

You are correct that if the sub-project believes its being exported as `p2` but the top-project tries to import it as `p1` and `p1 <> p2` then Coq refuses the .vo files as they `... contains library p2.<file> and not library p1.<file>`.

I would have no issue with dropping the logical path from -L as I can see it being quite confusing if you are unable to import due to names disagreeing, although maybe a more robust approach would be to allow `p2.<file>` to be imported as if it was `p1.<file>`?

silene commented:
> some sub-project depend on another sub-project that is not a strict sub-directory. (i.e. -L ../proj1 Proj1).
I believe that -R and -Q both require that any loaded packages be a sub-directory.

I am not sure it is much of an issue, since you also said that `-L` will act as `-R` after building. So, you will have to account for the fact that it might not be a subdirectory anyway.

> I would have no issue with dropping the logical path from -L as I can see it being quite confusing if you are unable to import due to names disagreeing

More importantly, there is no reason for the directory containing the `.v` files and the directory containing `_CoqProject` to be the same. In particular, there is no reason for the target `_CoqProject` file to be of the form `-R . Foo`. For example, all the libraries I maintain have `-R src Foo` in their `_CoqProject` file.

By the way, this invalidates my suggestion of using `-R` for that purpose.

Durbatuluk1701 commented:
So it seems that my initial crack at this still has design issues to be worked out. I will look into a new solution that hopefully addresses all the concerns.

Ultimately, I would like to make sure that the end goal of this modification seems palatable, a method for compiling "sub-projects" that can be imported and that "sub-projects" can exist on any path, not just sub-directories.

SkySkimmer commented:
>I believe that -R and -Q both require that any loaded packages be a sub-directory.

I don't think so?

Durbatuluk1701 commented:
> > I believe that -R and -Q both require that any loaded packages be a sub-directory.
> 
> I don't think so?

I say this because of the `check_overlapping_include` function in `tools/coq_makefile.ml` but I could be misconstruing its purpose

Zimmi48 commented:
> I was not sure if this change was major enough to merit a CEP, but I can try to write one up if we believe it is necessary.

It looks to me like the design discussion that just happened here reveals that a CEP would be useful here (although starting from a prototype is a good way of locating the places where the design could still be refined). The CEP doesn't have to be particularly long, but it can explain the feature both in terms of the user point of view (what will end up in the manual) and the design choices that it requires from the implementor point of view.

Durbatuluk1701 commented:
Understood, I will start work on writing up a CEP and add a link here to the CEP PR once ready

jfehrle commented on doc/sphinx/practical-tools/utilities.rst:
```diff
@@ -169,6 +169,24 @@ You can list multiple directories if you wish.
    a.v
    b.v
 
+If your project is made up of multiple sub-projects, you can specifically use the `-L`
```
This section is introducing basic concepts.  -L is not basic, it should be described later, perhaps in [this section](http://localhost:63342/coq2/_build/default/doc/refman-html/practical-tools/utilities.html#modifying-multiple-interdependent-projects-at-the-same-time)

IIUC you would not use -L to refer to other packages that you're not modifying as part of your project.

jfehrle commented:
@Durbatuluk1701 We need a clear explanation of how the feature will be used from the user's perspective and what it will let users do that they they can't do today (or is too awkward).

I don't see how -L adds very much for subprojects that are in subdirectories of another package--it seems like in most cases they can all be compiled using the top-level _CoqProject.  In the few cases where the parameters need to vary in the subproject, detecting _CoqProject in subdirectories without requiring -L is a good approach.  IIUC you wouldn't use -L to refer to installed packages, which as best practice, shouldn't be modified and thus won't need to be recompiled.

Using -L to refer to packages that are not in subdirectories is inherently machine-dependent, so it can't be used that way in published packages.

If there are multiple -L in the _CoqProject file, are they compiled in the order specified (i.e. user picks the order)?  (It would be nice if dependencies between subprojects could be discovered automatically and used to control the compilation order.)

If the number of subprojects is small, why can't the user just create a very short shell script that does a make on each of them rather than introduce -L?

Would subprojects work smoothly with git subproject checkout and Dune?

IMO the way we configure projects is already more complex for users than necessary.  There are a lot of quirks and caveats that are not easy to use or explain in the documentation.  Yet it is a topic that many users need to understand early.  The best improvements to configuration are ones that simplify the user experience and the explanation in the doc.

Thanks for proposing this.  I hope everyone's comments are helpful to you.  I look forward to hearing more about this.

gares commented:
At the beginning I thought this pr as a lightweight support for composition. But if has to become more complex, then I wonder how does it compare with dune, that supposedly supports this our of the box. The comparison should be part of the CEP.

tchajed commented:
Adding any feature to `_CoqProject` is non-trivial from an implementation perspective as it requires support from all IDEs and perhaps dune as well. This isn't a comment on this particular feature, just a note to keep in mind the complexity being added to the ecosystem; it's not enough to get `coq_makefile` working.
